rfcs/0049-flakes.md

```diff
@@ -0,0 +1,703 @@
+```
+inputs =
+  [ "nixpkgs"
+    github:edolstra/import-cargo
+  ];
+```
+then the resulting lock file might be:
+```
+{
+    "version": 2,
+    "inputs": {
+        "github:edolstra/import-cargo": {
+            "id": "import-cargo",
+            "inputs": {},
+            "narHash": "sha256-mxwKMDFOrhjrBQhIWwwm8mmEugyx/oVlvBH1CKxchlw=",
+            "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"
+        },
+        "nixpkgs": {
+            "id": "nixpkgs",
+            "inputs": {},
+            "narHash": "sha256-p7UqhvhwS5MZfqUbLbFm+nfG/SMJrgpNXxWpRMFif8c=",
+            "uri": "github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082"
+        }
+    }
+}
+```
+
+Thus, when we build this flake, the input `nixpkgs` is mapped to
+`github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e`. Nix
+will also check that the content hash of the input is equal to the one
+recorded in the lock file. This check is superfluous for Git
+repositories (since the commit hash serves a similar purpose), but for
+GitHub archives, we cannot directly check that the contents match the
+commit hash.
+
+Note that lock files are only used at top-level: the `flake.lock` files
+in dependencies (if they exist) are ignored.
+
+When you build a local repository (e.g. `nix build /path/to/repo`),
+Nix automatically creates a `flake.lock` file if it doesn't already
+exists. It will also update the lock file if inputs are added or
+removed. You can pass `--recreate-lock-file` to force Nix to recreate
+the lock file from scratch (and thus check for the latest version of
+each input).
+
+Lock files are not sufficient by themselves to ensure reproducible
+evaluation. It is also necessary to prevent certain impurities. In
+particular, the `nix` command now defaults to evaluating in "pure"
+mode, which means that the following are disallowed:
+
+* Access to files outside of the top-level flake or its inputs, as
+  well as paths fetched using `fetchTarball`, `fetchGit` and so on
+  without a commit hash or content hash. In particular this means that
+  Nixpkgs will not be able to use `~/.config/nixpkgs` anymore.
+
+* Access to the environment. This means that `builtins.getEnv "<var>"`
+  always returns an empty string.
+
+* Access to the system type (`builtins.currentSystem`).
+
+* Access to the current time (`builtins.currentTime`).
+
+* Use of the Nix search path (`<...>`); composition must be done
+  through flake inputs or `fetchX` builtins.
+
+Pure mode can be disabled by passing `--impure` on the command line.
+
+## Evaluation caching
+
+The fact that by default (i.e. in pure mode) evaluation is hermetic
+allows `nix` to cache flake attributes. For example (doing `nix build`
+on an already present package):
+
+    $ time nix build nixpkgs:firefox
+    real    0m1.497s
+
+    $ time nix build nixpkgs:firefox
+    real    0m0.052s
+
+The evaluation cache is kept in `~/.cache/nix/eval-cache-v1.sqlite`,
```

Comment:
Given that nix already has a mechanism for storing values identified by the hash of their inputs (i.e. /nix/store), it would seem conceptually neater to reuse that. It would also allow sharing some other utilities down the line (e.g. garbage collection of old entries) instead of reimplementing them.

It's possible that is impractical for performance reasons, but then maybe we can somehow speed up /nix/store for trivially small entries instead? Or put these in /nix/store as the source of truth, as well as in a sqlite cache on top of that.

Reply:
Using the Nix store would be orders of magnitude slower and take up way more disk space, so I don't think it's an option.